scripts/cluster-upgrade.sh

```diff
@@ -0,0 +1,138 @@
+#!/bin/bash
+
+OCP_VERSION_FROM=$1
+OCP_VERSION_TO=$2
+
+if [ -z $OCP_VERSION_TO ];
+then
+    echo "usage: $0 <from> <to>"
+    echo "example: $0 4.1.0 4.1.2"
+    exit 1
+fi
+
+UPGRADE_NOT_POSSIBLE=""
+UPGRADE_PROGRESSING=""
+UPGRADE_DONE=""
+
+NOW_EPOCH=$(date +"%s")
+
+# Verify we can actually upgrade (is this in the graph)
+CHANNEL_NAME=$(echo "${OCP_VERSION_TO}" | sed 's/\([^.]*\.[^.]*\)\..*/stable-\1/g')
+GRAPH=$(curl -s -H "Accept: application/json" https://api.openshift.com/api/upgrades_info/v1/graph?channel="${CHANNEL_NAME}")
+GRAPH_INDEX_FROM=$(echo "${GRAPH}" | jq -r "[ .nodes[] | .version == \"$OCP_VERSION_FROM\" ] | index(true)")
+GRAPH_INDEX_TO=$(echo "${GRAPH}" | jq -r "[ .nodes[] | .version == \"$OCP_VERSION_TO\" ] | index(true)")
+
+GRAPH_EDGE=$(echo "${GRAPH}" | jq -r ".edges[] | select(.[0] == $GRAPH_INDEX_FROM) | .[1] == $GRAPH_INDEX_TO" | grep true)
+
+if [ -z $GRAPH_INDEX_FROM ];
+then
+    echo "Cannot upgrade from $OCP_VERSION_FROM, it is not available in upgrade graph."
+    exit 2
+fi
+if [ -z $GRAPH_INDEX_TO ];
+then
+    echo "Cannot upgrade to $OCP_VERSION_TO, it is not available in upgrade graph."
+    exit 2
+fi
+if [ "$GRAPH_EDGE" != "true" ];
+then
+    echo "Cannot upgrade from $OCP_VERSION_FROM to $OCP_VERSION_TO, no path exists in upgrade graph."
+    exit 2
+fi
+
+# Verify the target version exists as a ClusterImageSet (CIS) on the cluster
+# (if not, the customer could never install to that target version)
+CIS_VERSION=$(oc get clusterimageset --all-namespaces -o json | jq -r ".items[].spec.releaseImage | split(\":\")[1] | select(. == \"$OCP_VERSION_TO\")")
+if [ "$OCP_VERSION_TO" != "$CIS_VERSION" ];
+then
+    echo "Cannot upgrade to $OCP_VERSION_TO, it is not available as a ClusterImageSet in the cluster."
+    exit 3
+fi
+
+# Verify we have ClusterDeployment CRs to work with
+if [ `oc get crd clusterdeployments.hive.openshift.io --no-headers 2>/dev/null | wc -l` == "0" ];
+then
+    echo "ERROR: Current cluster does not have Hive installed.  Verify where you are logged in."
+    exit 1
+fi
+
+TMP_DIR=`mktemp -d`
+
```

Comment:
```suggestion
if [[ $? -ne 0 ]];
then
    echo "ERROR: Couldn't create a temporary directory. Aborting.
    exit 1
fi
```

Add a protection for a failure to create a temporary directory. An alternative is to `set -e` and/or `set -o nounset` earlier in the program.